File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=s.length();
        unordered_set<char> set;
        int l=0,maxLen=0;
        for(int r=0;r<n;r++)
        {
            while(set.find(s[r])!=set.end())
            {
                set.erase(s[l]);
                l++;
            }
            set.insert(s[r]);
            maxLen=max(maxLen,(int)set.size());
        }
        return maxLen;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        Set<Character> st = new HashSet<>();
        int maxLength = 0;

        int l = 0;
        for(int r=0;r<n;r++) {
            char ch = s.charAt(r);
            while(st.contains(ch)) {
                st.remove(s.charAt(l));
                l++;
            }
            st.add(ch);
            maxLength = Math.max(maxLength,r-l+1);
        }
        return maxLength;
    }
}